Add star to linear pre-routing pass

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new transpiler pass `StarPreRouting` that adds a dedicated transformation that finds a star graph connectivity subcircuit and replaces it with a linear routing equivalent. This makes the circuit easier for a routing pass to work with.

### Details and comments

TODO:

- [x] Add more testing
- [x] Add final layout tracking and composition with routing.

Co-Authored-By: Sebastian Brandhofer <148463728+sbrandhsn@users.noreply.github.com>

mtreinish committed:
Add Star to linear pre-routing pass

This commit adds a new transpiler pass StarPreRouting that adds a
dedicated transformation that finds a star graph connectivity subcircuit
and replaces it with a linear routing equivalent. This makes the circuit
easier for a routing pass to work with.

mtreinish committed:
Merge branch 'main' into star-pre-routing

mtreinish committed:
Merge remote-tracking branch 'origin/main' into star-pre-routing

mtreinish committed:
Add missing doc imports

alexanderivrii commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,156 @@
+                    center_node = set(node.qargs)
+                    star_sequence.append(node)
+        if len(star_sequence) > 1 and center_node:
+            star_sequences.append((star_sequence, center_node))
```
I was thinking that we have multiple passes that collect blocks of nodes according to the following general strategy:
- if the current node matches a certain condition C1 and the current node can be added to the current block while satisfying a certain other condition C2, add the current node to the current block (and continue to the next node)
- else if the current block is empty, skip the current node (and continue to the next node)
- otherwise, add the current block to the list of collected blocks, set the current block to be empty, and repeat the first step with the empty current block

This is more general than what `BlockCollector` can currently do since it only collects nodes matches a given filter function (corresponding to C1), while now we want to say something like this: a node can be added to a partially constructed block either when the block is empty or one of the node's qubits is equal to the `center_qubit` (i.e. this is what the condition C2 would look like). 
If you think that having this more general set of block collection strategies would be useful, I will write a PR (I actually have code somewhere that does something similar already)

mtreinish replied:
There might be similar applications for block collection strategies like this, I'm not entirely sure though. But I can't imagine it would hurt to have the `BlockCollector` support this use case. I wouldn't mind having the block collection done in this pass be a bit more robust, it only works because of our default lexicographical topological sorting based on the qubit indices.

alexanderivrii commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,156 @@
+                            node.cargs,
+                        )
+                        processed_nodes.add(node)
+                    continue
```
Nice! Just to make sure that I understand this optimization: you are handling blocks of length 2 in a special manner since the qubits are already on a line, for example for `CX(0, 4) CX(0, 9)` we already have the line connectivity `4 -- 0 -- 9`, is this correct?

mtreinish replied:
Yeah, exactly if the block is of length 2 it will always be a line so there is no swapping needed.

alexanderivrii commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,156 @@
+                        )
+                        prev = inner_node.qargs
+                        processed_nodes.add(inner_node)
+                        continue
```
Oh, I see, not including the swap gate after the very first gate is ok because it's the previous optimization all over again. At first I was surprised that applying the pass to the circuit 
```
q_0: ──■────■────■────■──
     ┌─┴─┐  │    │    │  
q_1: ┤ X ├──┼────┼────┼──
     └───┘┌─┴─┐  │    │  
q_2: ─────┤ X ├──┼────┼──
          └───┘┌─┴─┐  │  
q_3: ──────────┤ X ├──┼──
               └───┘┌─┴─┐
q_4: ───────────────┤ X ├
                    └───┘
```
we get 
```
q_0: ──■────■───X─────────────────
     ┌─┴─┐  │   │                 
q_1: ┤ X ├──┼───┼─────────────────
     └───┘┌─┴─┐ │                 
q_2: ─────┤ X ├─X───■───X─────────
          └───┘   ┌─┴─┐ │         
q_3: ─────────────┤ X ├─X───■───X─
                  └───┘   ┌─┴─┐ │ 
q_4: ─────────────────────┤ X ├─X─
                          └───┘   
```
but I see now that the line consists of qubits `1 -- 0 -- 2 -- 3 -- 4`.

And you can drop the very last swap gate, since this already synthesizes the star up to a permutation gate, so the last swap can be absorbed into it.

alexanderivrii commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,156 @@
+                continue
+            else:
+                new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
+        return new_dag
```
To account for the induced permutation of qubits, would it make sense to add a `PermutationGate` at the end of the star rather than modifying the layout directly? And we already have the `ElidePermutations` pass that would be able to process these permutations.

mtreinish replied:
Well `ElidePermutations` hasn't merged yet, I'm still trying to figure out the layout issues there. But yeah we could in theory combine them, but yeah I like the idea of pairing them. I do wonder if there is a use case for this without adding the permutation, but doing this would work well. Maybe have it as a flag, like `reverse_permutation=True`, in the constructor to control whether we insert the `PermutationGate` or not?

alexanderivrii reviewed (commented):
I was playing a bit with this code, so decided to write a few high-level comments. 

This is a very cool pass that will greatly improve the quality of compiled circuits for various sparsely connected architectures.

The main question that I have (and this is more general than this PR) is that (if I understand correctly) this `star-to-line` pass will run before routing, with the intuition that for sparsely connected architectures we will find better layout + routing when the circuit consists of linear pieces rather than star pieces. If the circuit consists of a single star and the coupling graph contains a line of the right size, then VF2 will even find a perfect layout and no routing would be needed. However if there are two or more stars (or there are more 2-qubit gates), then even if the first star is mapped perfectly, Sabre might need to do many swaps before the second star, and it's not clear that the line connectivity of the second star will help. What do you think?

mtreinish commented:
> The main question that I have (and this is more general than this PR) is that (if I understand correctly) this star-to-line pass will run before routing, with the intuition that for sparsely connected architectures we will find better layout + routing when the circuit consists of linear pieces rather than star pieces. If the circuit consists of a single star and the coupling graph contains a line of the right size, then VF2 will even find a perfect layout and no routing would be needed. However if there are two or more stars (or there are more 2-qubit gates), then even if the first star is mapped perfectly, Sabre might need to do many swaps before the second star, and it's not clear that the line connectivity of the second star will help. What do you think?

I was thinking we would run this before the `layout` stage as part of `init` stage. In my local full path testing I was using the `pre_layout` stage but if we were to add it to the preset pass managers I'd use `init`. I called it prerouting but in my head I was thinking like you and we'd definitely want to do this before `VF2Layout` in case vf2 can find a line in the connectivity graph. Prerouting to me at the time I wrote this was a initial lighter routing pass, not intended to be right before the routing stage (like what was done in sabre pre-layout), but yeah it's a poor name choice as it's meaning is ambiguous and overloaded.

alexanderivrii commented:
There might be some problems with treating single-qubit gates. For example, the following circuit
```
     ┌───┐     ┌───┐     ┌───┐     ┌───┐     
q_0: ┤ H ├──■──┤ H ├──■──┤ H ├──■──┤ H ├──■──
     └───┘┌─┴─┐└───┘  │  └───┘  │  └───┘  │  
q_1: ─────┤ X ├───────┼─────────┼─────────┼──
          └───┘     ┌─┴─┐       │         │  
q_2: ───────────────┤ X ├───────┼─────────┼──
                    └───┘     ┌─┴─┐       │  
q_3: ─────────────────────────┤ X ├───────┼──
                              └───┘     ┌─┴─┐
q_4: ───────────────────────────────────┤ X ├
                                        └───┘
q_5: ────────────────────────────────────────                                          
```
gets transformed into
```
     ┌───┐             ┌───┐┌───┐┌───┐   
q_0: ┤ H ├──■────■───X─┤ H ├┤ H ├┤ H ├───
     └───┘┌─┴─┐  │   │ └───┘└───┘└───┘   
q_1: ─────┤ X ├──┼───┼───────────────────
          └───┘┌─┴─┐ │                   
q_2: ──────────┤ X ├─X───■────X──────────
               └───┘   ┌─┴─┐  │          
q_3: ──────────────────┤ X ├──X────■───X─
                       └───┘     ┌─┴─┐ │ 
q_4: ────────────────────────────┤ X ├─X─
                                 └───┘   
q_5: ────────────────────────────────────
```
Intuitively, this is because the pass only collects sequences of 2-qubit gates while ignoring 1-qubit gates (which is probably fine, since 1-qubit gates pose no problems for routing), but then forgets to take these 1-qubit gates into account (see how all of the Hadamard gates got pushed out to the end of the circuit).

alexanderivrii commented:
Another random thought is that star-to-line translation may create gate cancellation opportunities, i.e. we may want to rerun passes like inverse cancellation after star-to-line but before layout or routing.

mtreinish commented:
> Another random thought is that star-to-line translation may create gate cancellation opportunities, i.e. we may want to rerun passes like inverse cancellation after star-to-line but before layout or routing.

Yeah, that's a good point. I think when we're adding the pass to the preset pass managers we can determine the best place to add this. I'm wondering if we'll want to have a do while loop for initial optimization like we do for final optimization

mtreinish committed:
Merge remote-tracking branch 'origin/main' into star-pre-routing

mtreinish committed:
Fix handling of 1q gates and swap tracking

This commit fixes two related issues. The first is that it fixed the
handling of 1q gates in the middle of a star sequence. The sequence
collection was correctly skipping the 1q gates, but when we reassembled
the circuit and applied the linear routing it would incorrectly build
the routing without the 1q gates. This would result in the 1q gates
being pushed after the star. At the same time the other related issue
was that the permutations caused by the swap insertions were not taken
into account for subsequent gates. This would result in gates being
placed on the wrong bits and incorrect results being returned.

mtreinish commented:
> Intuitively, this is because the pass only collects sequences of 2-qubit gates while ignoring 1-qubit gates (which is probably fine, since 1-qubit gates pose no problems for routing), but then forgets to take these 1-qubit gates into account (see how all of the Hadamard gates got pushed out to the end of the circuit).

This was a good catch, I've pushed up a fix for this here: https://github.com/Qiskit/qiskit/pull/11387/commits/8cc0587a6aad7e7b1b9bfe92090643e6a517add6

sbrandhsn committed:
Merge branch 'main' into star-pre-routing

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=11387) <br/>All committers have signed the CLA.

sbrandhsn committed:
fixed/tested star-prerouting

sbrandhsn commented:
`StarPreRouting` works now but depends on #12057 and `FinalizeLayouts` in #9523 being merged (parts of this pass are adapted from #11614 ). If you have any suggestions/comments - I would love to hear them even before these two items are merged. :-)

Executed on a 6-qubit QFT quantum circuit, `StarPreRouting` yields the following mapped circuit:
```
                                                                          ░ ┌───┐                                                  ░ ┌───┐                                     ░ ┌───┐                         ░ ┌───┐          ░       ░ 
q_4 -> 0 ──────■──────────────────────────────────────────────────────────░─┤ H ├─■────────X───────────────────────────────────────░─┤ H ├─■────────X──────────────────────────░─┤ H ├─■────────X──────────────░─┤ H ├─■────────░───────░─
         ┌───┐ │P(π/2)                                                    ░ └───┘ │P(π/2)  │                                       ░ └───┘ │P(π/2)  │                          ░ └───┘ │P(π/2)  │              ░ └───┘ │P(π/2)  ░ ┌───┐ ░ 
q_5 -> 1 ┤ H ├─■────────■────────X────────────────────────────────────────░───────■────────X──■────────X───────────────────────────░───────■────────X──■────────X──────────────░───────■────────X──■────────X──░───────■────────░─┤ H ├─░─
         └───┘          │P(π/4)  │                                        ░                   │P(π/4)  │                           ░                   │P(π/4)  │              ░                   │P(π/4)  │  ░                ░ └───┘ ░ 
q_3 -> 2 ───────────────■────────X──■────────X────────────────────────────░───────────────────■────────X──■────────X───────────────░───────────────────■────────X──■────────X──░───────────────────■────────X──░────────────────░───────░─
                                    │P(π/8)  │                            ░                               │P(π/8)  │               ░                               │P(π/8)  │  ░                               ░                ░       ░ 
q_2 -> 3 ───────────────────────────■────────X──■─────────X───────────────░───────────────────────────────■────────X──■─────────X──░───────────────────────────────■────────X──░───────────────────────────────░────────────────░───────░─
                                                │P(π/16)  │               ░                                           │P(π/16)  │  ░                                           ░                               ░                ░       ░ 
q_1 -> 4 ───────────────────────────────────────■─────────X──■─────────X──░───────────────────────────────────────────■─────────X──░───────────────────────────────────────────░───────────────────────────────░────────────────░───────░─
                                                             │P(π/32)  │  ░                                                        ░                                           ░                               ░                ░       ░ 
q_0 -> 5 ────────────────────────────────────────────────────■─────────X──░────────────────────────────────────────────────────────░───────────────────────────────────────────░───────────────────────────────░────────────────░───────░─
                                                                          ░                                                        ░                                           ░                               ░                ░       ░ 
```
whereas `SabreLayout` yields:
```
                                                                    ░ ┌───┐                                               ░                                     ░ ┌───┐                         ░                   ░       ░ 
q_2 -> 0 ──────────────────X────────────────────────────────────────░─┤ H ├─X───────────■─────────────────────────────────░───────■─────────────────────────────░─┤ H ├─X───────────────────────░───────────────────░───────░─
                           │                                        ░ └───┘ │           │P(π/4)                           ░ ┌───┐ │P(π/2)                       ░ └───┘ │                       ░                   ░ ┌───┐ ░ 
q_4 -> 1 ──────■───────────X─────■──────────────────────────────────░───────X──■────────■────────X────────────────────────░─┤ H ├─■────────■───────────■────────░───────X──X───────────■────────░──────────■────────░─┤ H ├─░─
         ┌───┐ │P(π/2)           │P(π/8)                            ░          │P(π/2)           │                        ░ └───┘          │P(π/4)     │P(π/8)  ░          │           │P(π/4)  ░          │P(π/2)  ░ └───┘ ░ 
q_5 -> 2 ┤ H ├─■────────■────────■────────X─────────────────────────░──────────■─────────────────X──■────────X────────────░────────────────■────────X──■────────░──────────X──■────────■────────░───────X──■────────░───────░─
         └───┘          │P(π/4)           │                         ░                               │P(π/8)  │            ░                         │           ░             │P(π/2)           ░ ┌───┐ │           ░       ░ 
q_3 -> 3 ───────────────■─────────────────X──■─────────X────────────░───────────────────────────────■────────X──■─────────░───X─────────────────────X───────────░─────────────■─────────────────░─┤ H ├─X───────────░───────░─
                                             │P(π/16)  │            ░                                           │P(π/16)  ░   │                                 ░                               ░ └───┘             ░       ░ 
q_1 -> 4 ────────────────────────────────────■─────────X──■─────────░───X───────────────────────────────────────■─────────░───X─────────────────────────────────░───────────────────────────────░───────────────────░───────░─
                                                          │P(π/32)  ░   │                                                 ░                                     ░                               ░                   ░       ░ 
q_0 -> 5 ─────────────────────────────────────────────────■─────────░───X─────────────────────────────────────────────────░─────────────────────────────────────░───────────────────────────────░───────────────────░───────░─
                                                                    ░                                                     ░                                     ░                               ░                   ░       ░ 
```

A comparison on a 100-qubit QFT was done by running:
```
qc = qft(100)
cm = CouplingMap.from_line(100)

pm = generate_preset_pass_manager(optimization_level=3, coupling_map=cm, basis_gates=['rz', 'sx', 'cx'])
pm.pre_layout = PassManager([StarPreRouting()])
pm.post_scheduling = PassManager([FinalizeLayouts()])
start_ts = time()
result_star_prerouting = pm.run(qc)
end_ts = time()
print('star pre-routing done in {}s.'.format(end_ts-start_ts))
print('star pre-routing ops:', result_star_prerouting.count_ops())
print('star pre-routing depth:', result_star_prerouting.depth())

pm = generate_preset_pass_manager(optimization_level=3, coupling_map=cm, basis_gates=['rz', 'sx', 'cx'])
start_ts = time()
result_sabre = pm.run(qc)
end_ts = time()
print('sabre done in {}s.'.format(end_ts-start_ts))
print('sabre ops:', result_sabre.count_ops())
print('sabre depth:', result_sabre.depth())
```
This yielded the following data when barriers are included between the different 'stars' of a QFT instance:
```
star pre-routing done in 23.116374969482422s.
star pre-routing ops: OrderedDict([('cx', 14848), ('sx', 28635), ('rz', 26707), ('barrier', 100)])
star pre-routing depth: 39790

sabre done in 142.27110981941223s.
sabre ops: OrderedDict([('cx', 26635), ('sx', 6362), ('rz', 11522), ('barrier', 100)])
sabre depth: 27714
```
\- quite a larger decrease in `cx` gates! - and without barriers:
```
star pre-routing done in 23.750023126602173s.
star pre-routing ops: OrderedDict([('cx', 14848), ('sx', 20084), ('rz', 29150)])
star pre-routing depth: 1573

sabre done in 37.29286193847656s.
sabre ops: OrderedDict([('cx', 17510), ('sx', 4064), ('rz', 9848)])
sabre depth: 1764
```
which yields a 16% decrease in `cx` gates. These comparison are on a line graph.

mtreinish commented:
I've moved this out of draft since @sbrandhsn finished this up! But it still has dependencies on open PRs and will need to be rebased so I've marked this as on hold to indicate this.

mtreinish committed:
Merge branch 'main' into star-pre-routing

mtreinish commented on test/python/transpiler/test_star_prerouting.py:
```diff
@@ -0,0 +1,495 @@
+
+    # Skip level 3 because of unitary synth introducing non-clifford gates
+    @unittest.skipUnless(HAS_AER, "Aer required for clifford simulation")
+    @ddt.data(0, 1, 2)
```
We might need to exclude 2 here too since #12149 we're running unitary synthesis in opt level 2 too. I'll wait to see what CI says though.

sbrandhsn replied:
Yeah sounds good, let's see what CI tells us. :-)

sbrandhsn replied:
I ran CI locally to reduce the pressure on our merge-pipeline and noticed test fails so I excluded level 2, too.

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+        else:
+            return [
+                self.dag.get_node(pred_id) for pred_id in self.dag.direct_predecessors(node.node_id)
+            ]
```
I realize this was likely ported from #11614 but were you thinking we'd want to run this with a `DAGDependency`? I guess there is the potential to find a bigger star if we factor in potential commutations.

sbrandhsn replied:
Yup, @alexanderivrii realized we could run this pass while considering commutations in #11614 so I just left it in. Should I remove the paths corresponding to `DAGDependency`?

alexanderivrii replied:
I would suggest to remove the `DAGDependency` path for now, and this will get optimized automatically when we finalize #11614. Thanks!

sbrandhsn replied:
I don't think the paths on `DAGDependency` are fully functional as of now - at least I have not verified it...

mtreinish replied:
No it's fine, we don't have a great story about how to use `DAGDependency` via the pass managers yet. But I guess it's an option for someone running things manually and doesn't hurt at all.

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+            List[StarBlock]: a list of star blocks in the given dag
+            Union[List[DAGOpNode], List[DAGDepNode]]: a list of operations specifying processing order
+        """
+        self.dag = dag
```
I wouldn't use an instance variable for this. We'll keep around the output dag if we do this. We should either pass it directly to the methods that need the original dag or ensure we delete this prior to returning.

In general the pass will be run once but might live for a long time. By doing this we'll end up blocking python from freeing the memory during transpile because the refcount > 0 while the pass object exists.

sbrandhsn replied:
Thanks for noting this, I moved `self.dag` to a function parameter. :-)

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+            self.property_set["original_qubit_indices"] = input_qubit_mapping
+        self.property_set["virtual_permutation_layout"] = Layout(
+            {dag.qubits[out]: idx for idx, out in enumerate(qubit_mapping)}
+        )
```
What do we do here if `ElidePermutations` ran before this? Does it matter that we'll override the layout object previously set there?

alexanderivrii replied:
Right, we should compose the layouts. We should also compose layouts in ElidePermutations if we do not do this already.

jakelishman replied:
(fwiw this is why I was pushing hard for the formalisation of `final_permutation`, so we don't end up with things like some passes composing `virtual_permutation_layout`, some composing `final_layout`, and then `FinalizeLayouts` mixing the two together - we just have one well-defined responsibility for anything updating the same concept)

mtreinish replied:
I did this in #9523 with: https://github.com/Qiskit/qiskit/pull/9523/commits/28263bfdbc05c4a61b8fe60c9edf5fcaae0b87c3 we should do the same here too.

sbrandhsn replied:
I replicated `28263bf` in the new commit.

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+
+"""Search for star connectivity patterns and replace them with."""
+from typing import Iterable, Union, Optional, List, Tuple
+from numpy import floor, log10
```
I would use `math` for these since you're not running this on a numpy array. The overhead from `numpy` functions is typically higher because they do input checking.

```suggestion
from math import floor, log10
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+            else:
+                # the node is not part of a block
+                new_dag.apply_operation_back(
+                    node.op, _apply_mapping(node.qargs, qubit_mapping, dag.qubits), node.cargs
```
```suggestion
                    node.op, _apply_mapping(node.qargs, qubit_mapping, dag.qubits), node.cargs, check=False
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+                        new_dag.apply_operation_back(
+                            SwapGate(),
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
```
```suggestion
                            inner_node.cargs,
                            check=False,
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+                    new_dag.apply_operation_back(
+                        inner_node.op,
+                        _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                        inner_node.cargs,
```
```suggestion
                        inner_node.cargs,
                        check=False,
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+                        new_dag.apply_operation_back(
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
```
```suggestion
                            inner_node.cargs,
                            check=False,
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+                        new_dag.apply_operation_back(
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
```
```suggestion
                            inner_node.cargs,
                            check=False,
```

mtreinish commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -0,0 +1,409 @@
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
+                        )
```
We can avoid the extra runtime checking here because we know the input is valid:

```suggestion
                        new_dag.apply_operation_back(
                            inner_node.op,
                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
                            inner_node.cargs,
                            check=False,
                        )
```

mtreinish reviewed (commented):
Overall this looks great to me, thanks @sbrandhsn for taking this over and getting it across the finish line! I left a few inline comments, most of them are just questions or pretty minor.

mtreinish committed:
Apply suggestions from code review

alexanderivrii commented:
Thanks @mtreinish and @sbrandhsn for getting this completed, this is definitely going to be a super-useful pass. And thanks @jakelishman for doing this review (somehow I forgot that I was assigned to review this PR). I also looked at the code, and it looks great. In the future, we should probably refactor the star collection part of the pass into the more general block collection (i.e. finalize #11614) and replace the `virtual_permutation_layout`/`FinalizeLayouts` by `final_permutation`.

sbrandhsn committed:
changed moved self.dag to a function parameter

sbrandhsn committed:
Removing the usage of FinalizeLayout in star prerouting tests

sbrandhsn committed:
Merge branch 'main' into star-pre-routing

sbrandhsn committed:
format

sbrandhsn commented:
> Thanks @mtreinish and @sbrandhsn for getting this completed, this is definitely going to be a super-useful pass. And thanks @jakelishman for doing this review (somehow I forgot that I was assigned to review this PR). I also looked at the code, and it looks great. In the future, we should probably refactor the star collection part of the pass into the more general block collection (i.e. finalize #11614) and replace the `virtual_permutation_layout`/`FinalizeLayouts` by `final_permutation`.

Makes sense, thanks Jake, Matthew and Sasha for reviewing this! :-)

alexanderivrii reviewed (approved):
Thanks for all the last-minute fixes. LGTM!
